### DISCUSSION TITLE
Named client

### DISCUSSION THREAD

**Discussion by Arthur-Lee** [reactiveui/refit#1636] [General]
Hi, I just found the PR https://github.com/reactiveui/refit/pull/1418 which has already added named client. However I don't find the IRefitHttpClientFactory in the main branch.

Is there any plan to add this feature back? The named client can be very useful in some case.

**Comment by anaisbetts** [comment]
Is this covered by https://github.com/reactiveui/refit/blob/e5bc249551f85186235e6221be25962d1df53f65/README.md?plain=1#L1185

**Reply by Arthur-Lee** [reply]
I don't think the custom HttpClient can help in this scenario. Let me be more specific and take an example.

1. If we have two downstream service, they share same OpenAPI schema.
2. The middleware needs call the two downstream service by conditions. 
3. If there is only one downstream service, it will be fine. I can create ISomeApi, and bind the URL.

However the two downstream service share same Http schema. It seems that I can't inject two ISomeApi object with different name, and retrieve the ISomeApi object by the name just like HttpClient

```c#
builder.Services.AddHttpClient<SimpleClient>("one", client => client.BaseAddress = new("http://url_one"));

builder.Services.AddHttpClient<SimpleClient>("two", client => client.BaseAddress = new("http://url_two"));

IHttpClientFactory httpClientFactory;
var client = (SimpleClient)httpClientFactory.CreateClient("one");
```

**Comment by glennawatson** [comment]
The standalone IRefitHttpClientFactory from PR #1418 was not kept, but the exact scenario you describe (two instances of the same interface, different base URLs, retrieved by name) is now supported directly through keyed DI: AddKeyedRefitClient<T>(serviceKey, ...).

```csharp
builder.Services.AddKeyedRefitClient<ISomeApi>("one")
    .ConfigureHttpClient(c => c.BaseAddress = new Uri("http://url_one"));

builder.Services.AddKeyedRefitClient<ISomeApi>("two")
    .ConfigureHttpClient(c => c.BaseAddress = new Uri("http://url_two"));
```

Resolve by key:

```csharp
public class Middleware(
    [FromKeyedServices("one")] ISomeApi apiOne,
    [FromKeyedServices("two")] ISomeApi apiTwo) { ... }

// or
var apiOne = serviceProvider.GetRequiredKeyedService<ISomeApi>("one");
```

This is the Refit equivalent of resolving a named HttpClient by key, and it works when the interface/schema is shared across multiple downstream services. It requires the keyed-services support in .NET 8+ DI. Available in Refit.HttpClientFactory (see AddKeyedRefitClient overloads).
